File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    private int m;
    private int n;
    
    public int numIslands(char[][] grid) {
        if(grid.length == 0){
            return 0;
        }
        m = grid.length;
        n = grid[0].length;
        int ans = 0;
        
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j ++){
                if(grid[i][j]=='1'){
                    ans += 1;
                    dfs(grid, i, j);
                }
            }
        }
        
        return ans;
        
        
    }
    
    private void dfs(char[][] grid, int i, int j){
        boolean condition = i >= 0 && i < m && j >= 0 && j < n && grid[i][j]=='1';
        
        if(condition){
            grid[i][j] = '0';
            dfs(grid, i+1, j);
            dfs(grid, i-1, j);
            dfs(grid, i, j+1);
            dfs(grid, i, j-1);
        }
        
    }
}

File: 0200-number-of-islands/0200-number-of-islands.cs
public class Solution {
    public int NumIslands(char[][] grid) {
        int res = 0;
        for(int i = 0; i < grid.Length; i++){
            for(int j = 0; j < grid[0].Length; j++){
                if(grid[i][j] == '1'){
                    res += 1;
                    Helper(i, j, grid);
                }
            }
        }
        
        return res;
    }
    
    private void Helper(int i, int j, char[][]grid){
        if(grid[i][j] == '1'){
            grid[i][j] = '0';
        
            if(i + 1 < grid.Length){
                Helper(i + 1, j, grid);
            }

            if(i - 1 >= 0){
                Helper(i - 1, j, grid);
            }

            if(j + 1 < grid[0].Length){
                Helper(i, j + 1, grid);
            }

            if(j - 1 >= 0){
                Helper(i, j - 1, grid);
            }
        }
    }
    
}